src/denoising_data.jl

```diff
@@ -0,0 +1,105 @@
+export generate_HW
+
+# Function to unpad an array
+function unpad(x, n_p, m_p, n)
+  a = n_p - n
+  x = reshape_array(x, (n_p, m_p))
+  unpadded_x = x[(a + 1):end, (a + 1):end]
+  return unpadded_x
+end
+
+# Function to pad an array
+function pad(z, s1, s2)
+  x = cat(zeros(size(z, 1), s1), z, dims = 2)
+  x = cat(x, zeros(size(z, 1), s2), dims = 2)
+  x = cat(x, zeros(s2, size(x, 2)), dims = 1)
+  x = cat(zeros(s1, size(x, 2)), x, dims = 1)
+  return x
+end
+
+# Function to pad a specific array to match suitable dimensions
+function pad_x(z, a)
+  t = cat(zeros(size(z, 1), a), z, dims = 2)
+  t = cat(zeros(a, size(z, 2) + a), t, dims = 1)
+  return t
+end
+
+# Function to create a meshgrid for the kernel
+function meshgrid(x_range, y_range)
+  m = length(x_range)
+  n = length(y_range)
+  x = repeat(x_range, inner = (1, n))
+  y = repeat(y_range', outer = (m, 1))
+  return x, y
+end
+
+# Function to generate a Gaussian kernel
+function my_gaussian_kernel(kernel_size, kernel_sigma)
+  x, y = meshgrid((-kernel_size):kernel_size, (-kernel_size):kernel_size)
+  normal = 1 / (2.0 * pi * kernel_sigma^2)
+  kernel = exp.(-((x .^ 2 + y .^ 2) / (2.0 * kernel_sigma^2))) * normal
+  kernel = kernel / sum(kernel)
+  return kernel
+end
+
+# Function to generate a uniform kernel
+function uniform_kernel(kernel_size)
+  kernel = ones(2 * kernel_size + 1, 2 * kernel_size + 1)
+  kernel = kernel / sum(kernel)
+  return kernel
+end
+
+# Main function to generate H, H_T, W, and W_T functions
+function generate_HW(shape, shape_p, KERNEL_SIZE, KERNEL_TYPE, KERNEL_SIGMA = 1.5)
```

Comment:
This is not very good practice because not all arguments are used by all kernels. It would be better to have one function per kernel type.

Reply:
In fact, most papers don't use uniform kernel hence I can undo it

Reply:
Ok